src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs

```diff
@@ -12,34 +12,73 @@ namespace Microsoft.PowerShell.Commands
     [OutputType(typeof(System.IO.FileInfo))]
     public class NewTemporaryFileCommand : Cmdlet
     {
+        private string _extension = ".tmp";
+
+        /// <summary>
+        /// Specify a different file extension other than the default one, which is '.tmp'. The period in this parameter is optional
+        /// </summary>
+        [Parameter(Position = 0)]
+        [ValidateNotNullOrEmpty]
+        public string Extension
+        {
+            get { return _extension; }
+            set { _extension = value; }
+        }
+
         /// <summary>
         /// returns a TemporaryFile
         /// </summary>
         protected override void EndProcessing()
         {
             string filePath = null;
-            string tempPath = System.Environment.GetEnvironmentVariable("TEMP");
+            string tempPath = Path.GetTempPath();
+
             if (ShouldProcess(tempPath))
             {
-                try
+                int attempts = 0;
+                bool creationOfFileSuccessful = false;
+
+                // In case the random temporary file already exists, retry 
```

Comment:
The odds are low but we could make them really low by using e.g. Guids for the file name instead of `Path.GetRandomFileName()` because the latter is only 8 characters long...